### PR TITLE
Backend stores room changes instead of room state

### PR DESCRIPTION
**Pull request by sjohnson31**

### CODE REVIEW

File: api/src/game_state_server.py [hunk 1]
```diff
@@ -156,93 +104,118 @@ async def handle_connection(
...
                 with self.apm_transaction('connect'):
-                    result = await self.room_store.apply_mutation(
-                        room_id,
-                        request_id=None,
-                        mutate=lambda entities: self._acquire_room_slot(
-                            room_id, entities, client_ip
-                        ),
+                    if not self._room_context_by_id.get(room_id):
+                        await self._acquire_room_slot(room_id, client_ip)
+                        room = create_room(room_id, await self.room_store.read(room_id))
```
Comment (sburba) [inline api/src/game_state_server.py:112]: We need to ensure that we've successfully subscribed to changes before we read the room so we don't miss updates

File: api/src/game_state_server.py [hunk 2]
```diff
@@ -156,93 +104,118 @@ async def handle_connection(
...
+                        self._process_requests(room_id, requests, stop_fut)
+                    )
+                    async for msg in items_until(listener_q, stop_fut):
+                        yield msg
+                finally:
+                    request_task.cancel()
+                    self._room_occupancy[room_id] -= 1
+                    if self._room_occupancy[room_id] <= 0:
+                        del self._room_occupancy[room_id]
+                        self._room_context_by_id[room_id].change_listener.task.cancel()
```
Comment (sburba) [inline api/src/game_state_server.py:144]: We also want to delete the room if there isn't anyone left right?

File: api/src/game_state_server.py [hunk 3]
```diff
@@ -156,93 +104,118 @@ async def handle_connection(
...
-            with timber.context(request={'request_id': request.request_id}):
-                async for message in self._handle_request(
-                    updates=request.updates,
-                    room_id=room_id,
-                    request_id=request.request_id,
-                ):
-                    yield message
-
-    async def _handle_request(
+                self._room_occupancy[room_id] += 1
```
Comment (sburba) [inline api/src/game_state_server.py:131]: I think room_occupancy is the same as the number of queues in the change listener's `output_queues`

File: api/tests/test_room_store.py [hunk 4]
```diff
@@ -91,113 +83,27 @@ def any_room_store(func: Callable) -> Callable:
...
-    await room_store.apply_mutation(
-        'room-id-2', 'request-id-last', lambda _: mutate_to([])
-    )
+    await room_store.add_update('room-id-1', VALID_REQUEST)
+    await room_store.add_update('room-id-2', VALID_REQUEST)
 
     assert (await async_collect(room_store.get_all_room_ids())) == [
         'room-id-1',
         'room-id-2',
     ]
```
Comment (sburba) [inline api/tests/test_room_store.py:101]: Awesome how many tests we got to remove here

**Review by sburba — CHANGES_REQUESTED** [review]
Great work! Just some minor comments.

File: api/src/game_state_server.py (hunk 1 above)
Comment (sjohnson31) [reply api/src/game_state_server.py:112]: Good catch.

File: api/src/game_state_server.py (hunk 2 above)
Comment (sjohnson31) [reply api/src/game_state_server.py:144]: Probably.

File: api/src/game_state_server.py (hunk 3 above)
Comment (sjohnson31) [reply api/src/game_state_server.py:131]: Works for me.

File: api/src/api/api_structures.py [hunk 5]
```diff
@@ -29,7 +46,10 @@ class PingAction:
     data: Ping
 
 
+Update = Union[CreateOrUpdateAction, DeleteAction, PingAction]
```
[the remark this answers is not in the document]
Comment (sjohnson31) [reply api/src/api/api_structures.py:49]: Sure.

File: api/src/room_store/redis_room_store.py [hunk 6]
```diff
@@ -76,81 +45,32 @@ def _channel_key(room_id: str) -> str:
 
 
 @dataclass
-class ChangeListener:
-    output_queues: List[asyncio.Queue[RoomChangeEvent]]
-    task: asyncio.Task
+class SubscriptionResult:
```
[the remark this answers is not in the document]
Comment (sjohnson31) [reply api/src/room_store/redis_room_store.py:48]: You're right.

File: api/src/room_store/room_store.py [hunk 7]
```diff
@@ -44,34 +33,18 @@ class RoomChangeEvent:
...
-        room_id: str,
-        request_id: Optional[str],
-        mutate: Callable[
-            [Optional[EntityList]],
-            Union[
-                Awaitable[MutationResultType], Coroutine[Any, Any, MutationResultType]
-            ],
-        ],
-    ) -> MutationResultType:
+    async def add_update(self, room_id: str, request: Request) -> None:
```
[the remark this answers is not in the document]
Comment (sjohnson31) [reply api/src/room_store/room_store.py:45]: Sure.

File: api/src/util/async_util.py [hunk 8]
```diff
@@ -15,6 +15,19 @@ async def anext(iterator: AsyncIterator[_T]) -> _T:
     return await iterator.__anext__()
 
 
+async def items_until(it: asyncio.Queue[_T], stop: asyncio.Future) -> AsyncIterator[_T]:
```
[the remark this answers is not in the document]
Comment (sjohnson31) [reply api/src/util/async_util.py:18]: Sure.

**Review by sburba — APPROVED** [review]

**Pull request merged**